File: internal/repository/customer_repo.go
package repository

import (
	"Rocket-Invo/internal/database"
	"Rocket-Invo/internal/models"
	"time"
)

type CustomerRepo struct {
	db *database.DB
}

func NewCustomerRepo(db *database.DB) *CustomerRepo {
	return &CustomerRepo{db}
}

func (r *CustomerRepo) Create(c *models.Customer) (int64, error) {
	res, err := r.db.Exec(`
		INSERT INTO customers (name, phone, email, address, created_at)
		VALUES (?, ?, ?, ?, ?)
	`, c.Name, c.Phone, c.Email, c.Address, time.Now())
	if err != nil {
		return 0, err
	}
	return res.LastInsertId()
}

func (r *CustomerRepo) GetByID(id int64) (*models.Customer, error) {
	c := &models.Customer{}
	err := r.db.QueryRow(`
		SELECT id, name, phone, email, address, created_at
		FROM customers WHERE id = ?
	`, id).Scan(&c.ID, &c.Name, &c.Phone, &c.Email, &c.Address, &c.CreatedAt)
	if err != nil {
		return nil, err
	}
	return c, nil
}

func (r *CustomerRepo) GetAll() ([]models.Customer, error) {
	rows, err := r.db.Query(`
		SELECT id, name, phone, email, address, created_at
		FROM customers ORDER BY name
	`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var customers []models.Customer
	for rows.Next() {
		var c models.Customer
		if err := rows.Scan(&c.ID, &c.Name, &c.Phone, &c.Email, &c.Address, &c.CreatedAt); err != nil {
			return nil, err
		}
		customers = append(customers, c)
	}
	return customers, rows.Err()
}

func (r *CustomerRepo) Update(c *models.Customer) error {
	_, err := r.db.Exec(`
		UPDATE customers SET name=?, phone=?, email=?, address=? WHERE id=?
	`, c.Name, c.Phone, c.Email, c.Address, c.ID)
	return err
}

func (r *CustomerRepo) Delete(id int64) error {
	_, err := r.db.Exec("DELETE FROM customers WHERE id=?", id)
	return err
}

File: electron/database/customer-repo.ts
import { Database as SqlJsDatabase } from 'sql.js'
import { Customer } from '../types'
import { all, get, run, lastInsertId } from './query'

export class CustomerRepo {
  private db: SqlJsDatabase

  constructor(db: SqlJsDatabase) {
    this.db = db
  }

  getAll(): Customer[] {
    return all<Customer>(this.db, 'SELECT * FROM customers ORDER BY name')
  }

  getById(id: number): Customer | undefined {
    return get<Customer>(this.db, 'SELECT * FROM customers WHERE id = ?', [id])
  }

  create(data: Omit<Customer, 'id' | 'created_at'>): number {
    run(this.db, 'INSERT INTO customers (name, phone, email, address) VALUES (?, ?, ?, ?)',
      [data.name, data.phone, data.email, data.address])
    return lastInsertId(this.db)
  }

  update(data: Customer): void {
    run(this.db, 'UPDATE customers SET name = ?, phone = ?, email = ?, address = ? WHERE id = ?',
      [data.name, data.phone, data.email, data.address, data.id])
  }

  delete(id: number): void {
    run(this.db, 'DELETE FROM customers WHERE id = ?', [id])
  }
}
